### DISCUSSION TITLE
Duplicate vertices in the vertex shader?

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#558] [General]
I am attempting to make Inverted Hull lines in the shader as an experiment. I see that there is a vertex shader input and output, and that I can move this around, perform displace, etc. But is it possible to duplicate some vertices? There's no node, but I assume with Inline Code or writing my own node this would be possible? In geonodes, the process would be to duplicate the geometry, displace it, flip the normals, and then join it back with the original.

![image](https://github.com/bnpr/Malt/assets/36803956/66313041-ac91-470f-9ba8-5349b75cdc58)

I'm not clear on if we even can duplicate vertices in a vertex shader. I see a bunch of other contexts saying that you can't, and that you use other sorts of geometry shaders for this. Other places seem to imply that you can, but they may be using the term more generally.

**Comment by pragma37** [comment]
> I'm not clear on if we even can duplicate vertices in a vertex shader. I see a bunch of other contexts saying that you can't, and that you use other sorts of geometry shaders for this. Other places seem to imply that you can, but they may be using the term more generally.

Yeah, you can't do this in a vertex shader, you need geometry shaders for that, which Malt doesn't support.
Keep in mind also that geometry shaders (like tesselation shaders) tend to have quite bad performance and are pretty much considered a "bad practice".

So for this I would just duplicate the geometry in GN (or even render some objects twice using Render Nodes if you feel more adventurous), and then do the displacement or whatever in the shader.

**Reply by aVersionOfReality** [reply]
Ah okay, that makes sense.
By Render Nodes do you mean something like making a new pass or Render layer that renders the displaced hull, and then composite them together?

**Reply by pragma37** [reply]
Yes, I was going to put together an example, but then I realized that, while possible, there were a few missing pieces of the puzzle to make it nicer.
So I did a plugin instead:
[InvertedHulls.zip](https://github.com/user-attachments/files/15522502/InvertedHulls.zip)
![imagen](https://github.com/bnpr/Malt/assets/36610999/3a8f93f7-9202-444e-8111-8da10aba46ca)

**Reply by aVersionOfReality** [reply]
Oh that's neat. It took me a bit to figure out it was the Collection name it was using. That's a good idea!
Thanks for this example. I didn't realize we could use Python in Malt like this!
